include/uapi/sound/sof/header.h

```diff
@@ -22,7 +22,25 @@ struct sof_abi_hdr {
 	__u32 type;		/**< component specific type */
 	__u32 size;		/**< size in bytes of data excl. this struct */
 	__u32 abi;		/**< SOF ABI version */
-	__u32 reserved[4];	/**< reserved for future use */
+	union {
```

Comment:
@libinyang what's the use of the union here. What I was suggesting was to leave the reserved as is and move all ipc4 specific fields into the data with an addition ipc4 header

Reply:
It makes absolutely no sense to me to have a union between "something" and "reserved" stuff. This is unmanageable and opens the door to all kinds to issues with backwards compatibility.

I stop the review here.

Reply:
> @ranj063, by definition the `data` part is a black box and it should not be interpreted by the kernel at all, passed as it is. I don't think it is a good idea to start doing that from the kernel pow and from the tooling pow either.

@ujfalusi the whole thing including the sof_abi_hdr is "data" as faras the kernel is concerned. If we can parse that why not the .data ?

Reply:
Does the kernel do anything with the header + data except for forwarding to FW i.e. do we look at the header for size ?

Reply:
I really think we should leave sof_abi_hdr as is and create a new struct sof_ipc4_abi_hdr. @ujfalusi @plbossart what do you think?

Reply:
So, if the unionizing is not allowed, then the only reasonable option is to have a new header struct for tplg2/IPC4.
`sof_abi4_hdr` or `sof_ipc4_abi_hdr`, whatever.
But in that case the header does not need to copy the `sof_abi_hdr` layout, we must have the magic at the start, but it can be even 'S', 'O', 'F', '4', '\0' while we are at it, no?

With this we can consolidate and define the members of this header as we see appropriate, use **named** members instead of magic fields, like `extension`.

The other thing is the ipc_type (the what type of message should be used for the given blob). Is this depends on the
1. module: the given module always receives it's blob via the same message
2. blob type: a given type of blob is always sent via the same message
 2.1. can a module require different blob types? Isn't the blob type and a module is somehow locked?
3. mix of this two and the needed message type is really something which is just random?

if 1, then we could have a token in tplg to tell this? if it is 3, then we need to have the message type in the blob's header.

Reply:
Yes, I agree. A new struct sof_ipc4_abi_hdr should be a good solution. Or we can use below solution?
```
struct sof_abi_hdr {
        __u32 magic;
        __u32 type;             /**< component specific type */
        __u32 size;             /**< size in bytes of data excl. this struct */
        __u32 abi;              /**< SOF ABI version */
-       __u32 reserved[4];      /**< reserved for future use */
+      __u32 ipc_blob_type;
+      __u32 ipc_extension;
+      __u32 reserved[2];      /**< reserved for future use */
        __u32 data[];           /**< Component data - opaque to core */
 }  __packed;
```

Reply:
@libinyang new structure is the best way. and we do not want ipc_blob_type or extension in the header. The extn has to be set in the driver. Whatever you need for it such as param ID should be a separate field. as for blob_type, you saw Peter comment right? Is there a possibility of supporting multiple IPCs for receving blob data?

Reply:
@ujfalusi @ranj063 
I will use the new abi header. And please see my inline comments for the opens
> but it can be even 'S', 'O', 'F', '4', '\0' while we are at it,

yes, I will use 'S', 'O', 'F', '4', '\0' 

> Are we ever going to use sof_abi_hdr.type for IPC4 and if not then why not re-use it for the same purpose as the ipc_blob_type

If we are using the new ABI header, the old `type` field is removed. So only ipc_blob_type exists in the header. 

> The other thing is the ipc_type (the what type of message should be used for the given blob). Is this depends on the
    1. module: the given module always receives it's blob via the same message
    2. blob type: a given type of blob is always sent via the same message
    2.1. can a module require different blob types? Isn't the blob type and a module is somehow locked?
    3. mix of this two and the needed message type is really something which is just random?

> Is there a possibility of supporting multiple IPCs for receving blob data?

Each module may have several kcontrols. Each kcontrol is only for one purpose. For example, WoV may have one kcontrol named "wov_init" for init_instance, and one kcontrol "wov_keyword" for the keyword setting. Then user space knows it should use "wov_keyword" kcontrol to send the keyword data (just like user space knows to use 'xxx Volume' to control the volume), not "wov_init" kcontrol. The message type is something like the ipc_blob_type, but we only care it is for large_config, config or init.

> With this we can consolidate and define the members of this header as we see appropriate, use named members instead of magic fields, like ```extension```
I prefer to use ```ipc_extension```. This is because:
1. ```ipc_extension``` can be used in large_config and config. And extension is divided into several part for both large_config and config. So we must define several variables for the ```ipc_extension``` like below:
```
__u16 data /* used for config */;
__u16 param_id; /* used for both config and large_config */
__u32 data_size /* used for large_config */
```
Instead, we can use ```ipc_extension``` to represent them all.
2. ```extension``` from kcontrol is opaque to driver. driver doesn't need to parse the extension to do different operations. Driver only copies the ```ipc_extension``` to msg->extension. If we split the ```ipc_extension```, driver needs extra workload to combine them again.

Reply:
@libinyang this is so confusing. the IPC extension is only 32-bits isnt it? How can you have 64-bits of information in the extension coming from the topology. But I have to insist, we cannot pass on IPC message extensions blindly in the driver. This has to be set programmatically with all the needed information. Its bad enough dealing with corrupt data. WIth bad IPC extn, we're only increasing the risk.

Reply:
@libinyang this is what I'd like to propose. @ujfalusi what do you think?
```
/*
 * Identifies data type, size and ABI.
 * Used by any bespoke component data structures or binary blobs.
 */
struct sof_ipc4_abi_hdr {
        __u32 magic;            /**< 'S', 'O', 'F', '4', '\0' */
        __u32 size;             /**< size in bytes of data excl. this struct */
        __u32 abi;              /**< SOF ABI version */
	__u32 param_id;	 /**< parameter ID */
	__u32 type;	 	 /**< type of blob: MODULE_INIT, CONFIG or LARGE_CONFIG */
        __u32 reserved[3];      /**< reserved for future use */
        __u32 data[];           /**< Component data - opaque to core */
}  __packed;
```

Reply:
what's the definition of param_id?

Reply:
its the parameter ID associated with the the Kcontrol data. For example a module could have 2 parameters for module conifguration  and model data. Just thinking out loud here.

Reply:
I was confused by the earlier comment
__u16 param_id; /* used for both config and large_config */
And now the type definition
__u32 type;	 	 /**< type of blob: MODULE_INIT, CONFIG or LARGE_CONFIG */

if param_id is a different concept now, it should be clarified IMHO

Reply:
does that look better now?

Reply:
I am not sure. If we want to update e.g. a bunch of parameters for an EQ, is this really the case that it can be done in all of MODULE_INIT, CONFIG or LARGE_CONFIG.

It was my understanding that some parameters are only expected in specific blob types, and that is completely module and implementation specific. I vaguely recall a thread with @libinyang on this.

Reply:
@plbossart I think if the module doesnt support updating a particular parameter in one of the 3 types, it should handle it in the firmware and throw an error. This would give the topology writer enough information to amend the topology accordingly no?

Reply:
right, but in that case your wording of "This is valid with all blob_type's" is not really correct. What you meant to say is that the validity of the parameter in specific blob types is implementation dependent.

Reply:
also the range and acceptable values for the parameter ID are implementation specific and opaque to the kernel.

Reply:
ok makes sense. I've altered the definition now. Hopefully it makes more sense.

Reply:
copying the suggested definition here for clarity

````c
/*
 * Identifies data type, size and ABI.
 * Used by any bespoke component data structures or binary blobs.
 */
struct sof_ipc4_abi_hdr {
        __u32 magic;            /**< 'S', 'O', 'F', '4', '\0' */
        __u32 size;             /**< size in bytes of data excl. this struct */
        __u32 abi;              /**< SOF ABI version */
	__u32 blob_type;	 	 /**< type of blob: MODULE_INIT, CONFIG or LARGE_CONFIG */
	__u32 param_id;	 /**< ID indicating which parameter to update with the new data. The validity of the param_id with blob_type is dependent on the module implementation */
        __u32 reserved[3];      /**< reserved for future use */
        __u32 data[];           /**< Module data - opaque to core */
}  __packed;
````

Reply:
@plbossart 's suggestion of ```struct sof_ipc4_abi_hdr``` is almost good to me, but I would like to move ```_u32 size``` after ```__u32 reserved[3]``` as @plbossart suggested in https://github.com/thesofproject/linux/pull/3918#discussion_r994736712 . But I still have concern why we should use param_id, not just use extension. ```params_id``` is also opaque to us, no clearer to us than ```extension```. If we are using ```params_id```, the kernel has extra work to generate the ```extension``` based on ```params_id```

Reply:
what exactly do we achieve by putting the size after the rsvd[3]?

Reply:
The size of data[]. Pierre required "then call it data_size or something and move it just above data[]."